python/private/pypi/uv_pip_compile.bzl

```diff
@@ -0,0 +1,101 @@
+# Copyright 2024 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"Rule for locking third-party dependencies with uv."
+
+script_template = """\
+{uv} pip compile \
+--python {python} \
+--python-platform windows \
+--python-version 3.9 \
+--no-strip-extras \
+--generate-hashes \
+--output-file - \
+{dependencies_file}
+"""
+
+def _uv_pip_compile(ctx):
+    info = ctx.toolchains["//python:uv_toolchain_type"].uvtoolchaininfo
+    uv = info.binary
+
+    python = ctx.toolchains["@bazel_tools//tools/python:toolchain_type"].py3_runtime.interpreter
+    dependencies_file = ctx.file.dependencies_file
+
+    # Option 1: Build action option.
+    # Not really appropriate, but it executes.
```

Comment:
Yeah, having the following structure could work:
* `uv_pip_compile(name="lock_file")` - creates a set of requirements file when locking.
* `write_source_file(name="lock_file.update")` - takes the output of the files and writes to the repo.
* `diff_test(name="lock_file.test")` can be automatically created as part of this.

One would still have to do `bazel run :lock_file.update` to update the lock file, so the outside user semantics is still the same. WDYT?

We would probably need to sprinkle some `nocache` or `local` and `manual` tags, because the thing might require building wheels, which is non-hermetic. On the other hand, being able to depend upon a CC toolchain within this rule and set up the environment for `uv` to build the wheels would be super interesting. I think I would be +1 for having it a build action, yes it's a different way from before hand, but it may be better?

However, I remembered that the problem is that you may want to pass extra args (e.g. to update a particular package or similar) and for that having the CLI as a build action is not useful. So what you want to do is:
```console
bazel run :lock_file.update -- --update-package pylint
```

The above would update the lock file and bump `pylint` dependency to the latest available. I am not sure we can do this if this is running as a build action.

Reply:
Thanks for all the input. Yes, Im not yet set on the [interface for this rule](https://github.com/bazelbuild/rules_python/pull/1989/files#r1650959406). It was just some minimal wiring to execute a toolchain while I did the plumbing. 

I think lets land the toolchain, then we can add the downstream rule(s) and work out the interface and experience we want.